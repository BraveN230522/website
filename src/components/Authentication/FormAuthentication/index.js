import React, { useState, useContext, useCallback, useRef } from 'react';
import { useForm } from 'react-hook-form';
import { useLazyQuery } from '@apollo/react-hooks';
import { get } from 'lodash';

import { LOGIN_WITH_USER_NAME } from '@graphql/queries/login';
import { REGISTER_USER } from '@graphql/queries/me';

import UserContext from '@utils/userContext';

import {
  FormSection,
  FormWrapper,
  SectionTitle,
  InputLayout,
  Input,
  SubmitWrapper,
  SubmitButton,
  ErrorMessage,
  StyleForgotPassword,
} from './style';

const FormAuthentication = ({ formTitle, submitButtonTitle, afterSubmit, onClick }) => {
  // TODO: add registration query as well

  const { dispatch: userDispatch } = useContext(UserContext);

  const [serverError, setServerError] = useState(null);

  const { register, handleSubmit, errors, watch } = useForm();

  const password = useRef({});
  password.current = watch('password', '');

  const handleServerErrors = error => {
    const errorMessage = error.graphQLErrors && error.graphQLErrors[0] ? error.graphQLErrors[0].message : error.message;
    setServerError(errorMessage);
  };

  const processLogin = loginData => {
    userDispatch({
      type: 'login',
      data: { ...loginData, id: loginData.wpid || loginData.id },
    });
    afterSubmit();
  };

  const onLoginSucess = response => {
    const loginDataResult = get(response, 'loginWithUserName', {});

    if (!loginDataResult || !loginDataResult.userToken) {
      return setServerError('Invalid username or password!');
    }

    window.lxt('identify', 'email', '{{Lexer - Login Email}}');
    window.lxt('converted');

    processLogin(loginDataResult);
  };

  const onRegisterSuccess = response => {
    const userData = get(response, 'register', {});

    if (!userData || !userData.wpid || !userData.userToken) {
      return setServerError('Something went wrong');
    }

    window.lxt('identify', 'email', '{{Lexer - Newsletter Email}}');
    window.lxt('converted');

    window.lxt('identify', 'email', '{{Lexer - Registration Email}}');
    window.lxt('converted');

    processLogin(userData);
  };

  const [registerUser, { loading: submittingRegister }] = useLazyQuery(REGISTER_USER, {
    onCompleted: onRegisterSuccess,
    onError: handleServerErrors,
    fetchPolicy: 'no-cache',
  });

  const [loginWithUserName, { loading: submitting }] = useLazyQuery(LOGIN_WITH_USER_NAME, {
    onCompleted: onLoginSucess,
    onError: handleServerErrors,
    fetchPolicy: 'no-cache',
  });

  const isRegistering = submitButtonTitle === 'Register';

  const onSubmit = data => {
    if (isRegistering) {
      registerUser({
        variables: {
          email: data.email,
          password: data.password,
          firstName: data.firstName,
          surname: data.lastName,
        },
      });
      return;
    }

    loginWithUserName({
      variables: {
        // email: data.email, // New API will use email (which it expects via the username parameter)
        username: data.email,
        password: data.password,
      },
    });
  };

  const onChange = useCallback(() => {
    setServerError(null);
  }, []);

  return (
    <FormSection data-testid="form-authentication-component">
      {formTitle && <SectionTitle>{formTitle}</SectionTitle>}

      <FormWrapper onSubmit={handleSubmit(onSubmit)}>
        <InputLayout isRegistering={isRegistering}>
          {isRegistering && (
            <>
              <Input
                name="email"
                placeholder="Email Address"
                type="text"
                register={register({
                  required: true,
                  pattern: /^\S+@\S+\.\S+$/i,
                })}
                error={errors.email && 'Provide a valid Email Address.'}
                ariaLabel="email-address-input"
                onChange={onChange}
              />
              <Input
                name="firstName"
                placeholder="First Name"
                type="text"
                register={register({
                  required: isRegistering,
                })}
                error={errors.firstName && 'First Name is required.'}
                ariaLabel="first-name-input"
                onChange={onChange}
              />
              <Input
                name="lastName"
                placeholder="Last Name"
                type="text"
                register={register({
                  required: isRegistering,
                })}
                error={errors.surname && 'Last Name is required.'}
                ariaLabel="last-name-input"
                onChange={onChange}
              />
            </>
          )}
          {/* New authentication API will use only email not email and username */}
          {!isRegistering && (
            <Input
              name="email"
              placeholder="Email Address"
              type="text"
              register={register({
                required: true,
                pattern: /^\S+@\S+\.\S+$/i,
              })}
              error={errors.email && 'Provide a valid Email Address.'}
              ariaLabel="email-address-input"
              onChange={onChange}
            />
          )}
          <Input
            name="password"
            placeholder="Password"
            type="password"
            register={register({
              required: 'Password required',
              minLength: {
                value: 8,
                message: 'Password must have at least 8 characters',
              },
            })}
            error={errors.password && errors.password.message}
            ariaLabel="password-input"
            onChange={onChange}
          />
          {!isRegistering && <StyleForgotPassword onClick={onClick}>Forgot password ?</StyleForgotPassword>}
          {isRegistering && (
            <Input
              name="confirmPassword"
              placeholder="Confirm password"
              type="password"
              register={register({
                required: 'You need to confirm the password',
                validate: value => value === password.current || 'The passwords do not match',
              })}
              error={errors.confirmPassword && errors.confirmPassword.message}
              ariaLabel="confirm-password-input"
              onChange={onChange}
            />
          )}
        </InputLayout>

        {serverError && <ErrorMessage>{serverError}</ErrorMessage>}

        <SubmitWrapper>
          <SubmitButton disabled={submitting || submittingRegister} variant="primary" type="submit">
            {submitting || submittingRegister ? 'Submitting' : submitButtonTitle}
          </SubmitButton>
        </SubmitWrapper>
      </FormWrapper>
    </FormSection>
  );
};

export default FormAuthentication;
