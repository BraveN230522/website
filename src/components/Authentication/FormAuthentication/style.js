import styled from '@emotion/styled';
import React from 'react';

import Button from '@components/common/Form/Button';

const FormSection = styled.section`
  display: flex;
  flex-direction: column;
`;

const FormWrapper = styled.form`
  display: flex;
  flex-direction: column;
  width: 100%;
`;

const InputLayout = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  width: 100%;

  div {
    width: ${({ isRegistering }) => (isRegistering ? 'calc(50% - 4px)' : '100%')};
  }

  ${({ theme }) => theme.media.phone`
    display: flex;
    flex-direction: column;
    flex-wrap: nowrap;
   div {
        width: 100%;
      }
    }
`};
`;

const InputWrapper = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 60px;
  ${({ isEmail }) => (isEmail ? 'width: 100% !important' : '')}
`;

const InputField = styled.input`
  display: flex;
  height: 50px;
  width: 100%;
  justify-content: center;
  align-items: center;
  font-family: ${({ theme }) => theme.fonts.Inter};
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  outline: none;

  padding: 12px 24px;
  margin: 0;
  margin-bottom: 10px;

  border: 2px solid ${({ theme }) => theme.colors.browGray};
  box-sizing: border-box;
  border-radius: 30px;

  ${({ theme }) => theme.media.phone`
    border-width: 1px;
    box-sizing: border-box;
    border-radius: 25px;
  `};
`;

const ErrorMessage = styled.span`
  display: flex;
  width: 100%;
  justify-content: center;
  align-items: center;
  font-family: ${({ theme }) => theme.fonts.Inter};
  font-style: normal;
  font-weight: 400;
  font-size: 14px;

  margin-bottom: -5px;
  margin-bottom: 20px;

  color: ${({ theme }) => theme.colors.error};
`;

const UserName = styled.p`
  font-size: 1.4rem;
`;

const Input = ({ name, placeholder, type, register, error, ariaLabel, onChange }) => {
  return (
    <InputWrapper isEmail={name === 'email'}>
      <InputField
        name={name}
        placeholder={placeholder}
        type={type}
        ref={register}
        aria-label={ariaLabel}
        onChange={onChange}
      />
      {error && <ErrorMessage>{error}</ErrorMessage>}
    </InputWrapper>
  );
};

const SectionTitle = styled.span`
  font-family: ${({ theme }) => theme.fonts.Inter}
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  text-align: center;

  height: 16px;
  padding: 0;
  margin: 0;
  margin-bottom: 12px;
  box-sizing: border-box;

  color: ${({ theme }) => theme.colors.black};

`;

const SubmitWrapper = styled.div`
  display: flex;
  width: 100%;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  padding-top: 24px;
  ${({ theme }) => theme.media.phone`
      padding-top: 20%;
  `};
`;

const SubmitButton = styled(Button)`
  width: 42%;

  ${({ theme }) => theme.media.phone`
      width: 100%;
  `};
`;

const WrapperButton = styled.div`
  display: flex;
  width: 250px;
  justify-content: space-between;
`;

const StyleForgotPassword = styled.div`
  display: flex;
  justify-content: flex-end;
  cursor: pointer;
  color: #2891ff;
`;

export {
  FormSection,
  UserName,
  FormWrapper,
  SectionTitle,
  InputLayout,
  Input,
  SubmitWrapper,
  SubmitButton,
  ErrorMessage,
  WrapperButton,
  StyleForgotPassword,
};
