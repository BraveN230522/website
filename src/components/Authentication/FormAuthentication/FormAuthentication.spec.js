import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import withGlobals from '@swiamTestUtils/withGlobals';
import { matchers } from 'jest-emotion';

import FormAuthentication from './index';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);

afterEach(() => {
  cleanup();
});

const testId = 'form-authentication-component';

const setup = (label, type) => {
  const utils = render(<FormAuthentication submitButtonTitle={type || 'Login'} />, {
    wrapper: withGlobals,
  });

  if (!label) {
    return utils;
  }

  let input;
  if (typeof label === 'string') {
    input = utils.getByLabelText(label);
  } else {
    input = label.map(_ => utils.getByLabelText(_));
  }

  return {
    input,
    ...utils,
  };
};

describe('<FormAuthentication />', () => {
  it('Renders the Authentication unchanged', () => {
    const { asFragment, queryByTestId } = setup();
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId(testId)).toBeTruthy();
  });

  it('The Login form can be filled', () => {
    const { input } = setup(['email-address-input', 'password-input'], 'Login');
    const [email, password] = input;

    fireEvent.change(email, {
      target: { value: 'james@bond.com' },
    });
    expect(email.value).toBe('james@bond.com');

    fireEvent.change(password, {
      target: { value: 'my-secret-123' },
    });
    expect(password.value).toBe('my-secret-123');
  });

  it('The Register form can be filled', () => {
    const { input } = setup(
      [
        'display-name-input',
        'first-name-input',
        'last-name-input',
        'email-address-input',
        'password-input',
        'confirm-password-input',
      ],
      'Register'
    );
    const [displayName, firstName, lastName, email, password, confirmPassword] = input;

    fireEvent.change(displayName, {
      target: { value: 'JamesBond' },
    });
    expect(displayName.value).toBe('JamesBond');

    fireEvent.change(firstName, {
      target: { value: 'James' },
    });
    expect(firstName.value).toBe('James');

    fireEvent.change(lastName, {
      target: { value: 'Bond' },
    });
    expect(lastName.value).toBe('Bond');

    fireEvent.change(email, {
      target: { value: 'james@bond.com' },
    });
    expect(email.value).toBe('james@bond.com');

    fireEvent.change(password, {
      target: { value: 'my-secret-123' },
    });
    expect(password.value).toBe('my-secret-123');

    fireEvent.change(confirmPassword, {
      target: { value: 'my-secret-1235555' },
    });
    expect(confirmPassword.value).toBe('my-secret-1235555');
  });
});
