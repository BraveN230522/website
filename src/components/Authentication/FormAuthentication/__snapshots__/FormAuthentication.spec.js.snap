// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormAuthentication /> Renders the Authentication unchanged 1`] = `
<DocumentFragment>
  .emotion-17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
}

.emotion-8 div {
  width: 100%;
}

@media (max-width:48em) {
  .emotion-8 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
  }

  .emotion-8 div {
    width: 100%;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 50px;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Inter,sans-serif;
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  padding: 12px 24px;
  margin: 0;
  margin-bottom: 10px;
  border: 2px solid #9c9c9c;
  box-sizing: border-box;
  border-radius: 30px;
}

@media (max-width:48em) {
  .emotion-0 {
    border-width: 1px;
    box-sizing: border-box;
    border-radius: 25px;
  }
}

.emotion-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 24px;
}

@media (max-width:48em) {
  .emotion-13 {
    padding-top: 20%;
  }
}

.emotion-11 {
  border: solid 2px #2891ff;
  color: #ffffff;
  background: #2891ff;
  height: 45px;
  font-weight: 600;
  font-size: 16px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 0.3s ease-in;
  transition: all 0.3s ease-in;
  cursor: pointer;
  font-family: Inter,sans-serif;
  border-radius: 25px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 103px;
  width: 33.33%;
}

.emotion-11:hover {
  border-color: #2891ff;
  background: #ffffff;
  color: #2891ff;
}

.emotion-11 svg {
  margin-right: 6px;
}

.emotion-11:hover path {
  stroke: #000000;
}

@media (max-width:48em) {
  .emotion-11 {
    width: 100%;
  }
}

<section
    class="emotion-17 emotion-18"
    data-testid="form-authentication-component"
  >
    <form
      class="emotion-15 emotion-16"
    >
      <div
        class="emotion-8 emotion-9"
      >
        <div
          class="emotion-2 emotion-3"
        >
          <input
            aria-label="email-address-input"
            class="emotion-0 emotion-1"
            name="email"
            placeholder="Email Address"
            type="text"
          />
        </div>
        <div
          class="emotion-2 emotion-3"
        >
          <input
            aria-label="password-input"
            class="emotion-0 emotion-1"
            name="password"
            placeholder="Password"
            type="password"
          />
        </div>
      </div>
      <div
        class="emotion-13 emotion-14"
      >
        <button
          class="emotion-10 emotion-11 emotion-12"
        >
          Login
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
